linux/LinuxProcessTable.c

```diff
@@ -78,11 +78,68 @@ static FILE* fopenat(openat_arg_t openatArg, const char* pathname, const char* m
    return fp;
 }
 
+static long fast_strtol_dec(char** str, int maxlen) {
+   long result = 0;
+   bool neg = false;
+
+   if (**str == '-') {
+      neg = true;
+      (*str)++;
+   }
+
+   if (!maxlen)
+      maxlen = 20; // −9223372036854775808
+
+   while (maxlen-- && **str >= '0' && **str <= '9') {
+      result *= 10;
+      result += **str - '0';
+      (*str)++;
+   }
+
+   return neg ? -result : result;
+}
+
+static unsigned long fast_strtoul_dec(char** str, int maxlen) {
+   unsigned long result = 0;
+
+   if (!maxlen)
+      maxlen = 20; // 18446744073709551615
+
+   while (maxlen-- && **str >= '0' && **str <= '9') {
+      result *= 10;
+      result += **str - '0';
+      (*str)++;
+   }
+
+   return result;
+}
+
+static long long fast_strtoll_dec(char** str, int maxlen) {
+   long long result = 0;
+   bool neg = false;
+
+   if (**str == '-') {
+      neg = true;
+      (*str)++;
+   }
+
+   if (!maxlen)
+      maxlen = 20; // −9223372036854775808
+
+   while (maxlen-- && **str >= '0' && **str <= '9') {
+      result *= 10;
+      result += **str - '0';
+      (*str)++;
+   }
+
+   return neg ? -result : result;
+}
+
 static inline uint64_t fast_strtoull_dec(char** str, int maxlen) {
    register uint64_t result = 0;
 
    if (!maxlen)
-      --maxlen;
+      maxlen = 20; // 18446744073709551615
```

Comment:
Documenting the previous default makes no sense for this line. If at all we want documentation it would explain, that this is "assume max reasonable length if none given". I'm for dropping this comment though …

Reply:
Just FYI, the standard `strtoull()` permits almost unlimited number of leading zeros in the string. While it won't certainly happen in parsing any Linux `procfs` file, having an arbitrary length limit in a `strtoull()` implementation would technically violate ISO C.

Reply:
The issue here is less about ISO C compliance, but more about squeeze out cycles. `fast_strtoull` is not meant to comply to ISO C, but to parse the data we actually encounter in realistic situations. Otherwise we'd have to support locales and a full load of other stuff (different radix prefixes) we don't need …